jpegenc/subblocks/dct_1d/dct_1d.py

```diff
@@ -0,0 +1,205 @@
+
+    mult_reg = [Signal(intbv(0, min=-mult_max_range, max=mult_max_range))
+                for _ in range(N)]
+    adder_reg = [Signal(intbv(0, min=-mult_max_range, max=mult_max_range))
+                 for _ in range(N)]
+    coeffs = [Signal(intbv(0, min=-coeff_range, max=coeff_range))
+              for _ in range(N)]
+    mux_flush = [Signal(intbv(0, min=-mult_max_range, max=mult_max_range))
+                 for _ in range(N)]
+
+    inputs_counter = Signal(intbv(0, min=0, max=N))
+    cycles_counter = Signal(intbv(0, min=0, max=N+4))
+    first_row_passed = Signal(bool(0))
+
+    data_in_reg = Signal(intbv(0, min=-2**nbits,
+                               max=2**nbits))
+    # coefficient rom
+    dct_1d_obj = dct_1d_transformation(N)
+    coeff_matrix = dct_1d_obj.dct_int_coeffs(fract_bits)
+    coeff_rom = tuple_construct(coeff_matrix)
+
+    output_sigs = [Signal(intbv(0, min=-2**output_fract, max=2**output_fract))
+                   for _ in range(N)]
+
+    @always_seq(clock.posedge, reset=reset)
+    def input_reg():
+        """input register"""
+        data_in_reg.next = input_interface.data_in
+
+    @always_seq(clock.posedge, reset=reset)
+    def coeff_assign():
+        """coefficient assignment from rom"""
+        if input_interface.data_valid:
+            for i in range(N):
+                coeffs[i].next = coeff_rom[i * N + int(inputs_counter)]
+
+    @always_seq(clock.posedge, reset=reset)
+    def mul_add():
+        """multiplication and addition"""
+        if input_interface.data_valid:
+            for i in range(N):
+                mult_reg[i].next = data_in_reg * coeffs[i]
+                adder_reg[i].next = mux_flush[i] + mult_reg[i]
+
+    @always_comb
+    def mux_after_adder_reg():
+        """after 8 inputs flush one of the inputs of the adder"""
+        if cycles_counter == (N + 2) or (cycles_counter == (N - 1)
+                                         and first_row_passed):
+            for i in range(N):
+                mux_flush[i].next = 0
+        else:
+            for i in range(N):
+                mux_flush[i].next = adder_reg[i]
+
+    @always_seq(clock.posedge, reset=reset)
+    def counters():
+        """inputs and cycles counter"""
+        if input_interface.data_valid:
+            if cycles_counter == (N + 2) or (cycles_counter == (N - 1)
+                                             and first_row_passed):
+                cycles_counter.next = 0
+                first_row_passed.next = True
+            else:
+                cycles_counter.next = cycles_counter + 1
+
+            if inputs_counter == N - 1:
+                inputs_counter.next = 0
+            else:
+                inputs_counter.next = inputs_counter + 1
+
+    @always_seq(clock.posedge, reset=reset)
+    def outputs():
+        """rounding"""
+        if cycles_counter == N + 2 or (first_row_passed and
+                                       cycles_counter == N - 1):
+            for i in range(N):
```

Comment:
Does the for loop need to be in the conditional?  I think only
the valid does, this should reduce the logic in the datapath.